### PR TITLE
deviceTree.applyOverlays: rewrite in Python/libfdt

### PR DESCRIPTION
**Pull request by K900**
## Description of changes

Pros:
- code you can actually read
- compatible strings handled properly
- no shelling out to what's essentially test tools

Cons:
- Python

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/os-specific/linux/device-tree/default.nix [hunk 1]
```diff
@@ -26,41 +26,11 @@
...
-          mv "$dtb"{,.in}
-          fdtoverlay -o "$dtb" -i "$dtb.in" "${o.dtboFile}"
-          echo "ok"
-          rm "$dtb.in"
-        fi
-        '')}
-
-      done
+    nativeBuildInputs = [
+      (python3.withPackages(ps: [ps.libfdt]))
```
Comment (Artturin) [inline pkgs/os-specific/linux/device-tree/default.nix:30]:
```suggestion
      (python3.pythonOnBuildForHost.withPackages(ps: [ps.libfdt]))
```

For cross

Comment (K900) [reply pkgs/os-specific/linux/device-tree/default.nix]: Applied, thanks

**Comment by K900** [comment]
OK, I guess as much testing as realistic has been done here, let's send it.

**Comment by flokli** [comment]
This regressed device tree overlays for me.

I'm using the following device tree overlay on a 6.11 kernel:

```
  hardware.deviceTree.overlays = [
    {
      name = "fan-cooling-levels";
      dtsText = ''
        #include <dt-bindings/gpio/gpio.h>
        #include <dt-bindings/interrupt-controller/irq.h>
        #include <dt-bindings/pinctrl/rockchip.h>
        #include <dt-bindings/thermal/thermal.h>

        /dts-v1/;
        /plugin/;
        / {
          compatible = "xunlong,orangepi-5-plus", "rockchip,rk3588";
        };

        &fan {
          cooling-levels = <0 65 70 80 100>;
        };

        &package_thermal {
          polling-delay = <1000>;

          trips {
            package_fan0: package-fan0 {
              temperature = <55000>;
              hysteresis = <2000>;
              type = "active";
            };

            package_fan1: package-fan1 {
              temperature = <65000>;
              hysteresis = <2000>;
              type = "active";
            };
          };

          cooling-maps {
            map1 {
              trip = <&package_fan0>;
              cooling-device = <&fan THERMAL_NO_LIMIT 1>;
            };

            map2 {
              trip = <&package_fan1>;
              cooling-device = <&fan 2 THERMAL_NO_LIMIT>;
            };
          };
        };
      '';
    }
  ];

```


With this PR, the build fails:

```
Processing source device tree rockchip/rk3588-armsom-sige7.dtb...
  Applying overlay fan-cooling-levels
Saving final device tree rockchip/rk3588-armsom-sige7.dtb...
Processing source device tree rockchip/rk3588-coolpi-cm5-evb.dtb...
  Applying overlay fan-cooling-levels
Traceback (most recent call last):
  File "/nix/store/206l37y6cvcwl3ck1ppwgnxn4cnm818i-apply_overlays.py", line 102, in <module>
    main()
  File "/nix/store/206l37y6cvcwl3ck1ppwgnxn4cnm818i-apply_overlays.py", line 92, in main
    dt = apply_overlay(dt, overlay.fdt)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/206l37y6cvcwl3ck1ppwgnxn4cnm818i-apply_overlays.py", line 50, in apply_overlay
    raise FdtException(err)
libfdt.FdtException: pylibfdt error -1: FDT_ERR_NOTFOUND
```

Without it (reverting this PR), it succeeds:

```
[…]
Skipping overlay fan-cooling-levels: incompatible with rk3588-evb1-v10.dtb
Skipping overlay fan-cooling-levels: incompatible with rk3588-friendlyelec-cm3588-nas.dtb
Skipping overlay fan-cooling-levels: incompatible with rk3588-jaguar.dtb
Skipping overlay fan-cooling-levels: incompatible with rk3588-nanopc-t6.dtb
Skipping overlay fan-cooling-levels: incompatible with rk3588-ok3588-c.dtb
Applying overlay fan-cooling-levels to rk3588-orangepi-5-plus.dtb... ok
Skipping overlay fan-cooling-levels: incompatible with rk3588-quartzpro64.dtb
Skipping overlay fan-cooling-levels: incompatible with rk3588-rock-5-itx.dtb
Skipping overlay fan-cooling-levels: incompatible with rk3588-rock-5b.dtb
Skipping overlay fan-cooling-levels: incompatible with rk3588-tiger-haikou.dtb
Skipping overlay fan-cooling-levels: incompatible with rk3588-toybrick-x0.dtb
[…]
```

It seems it tries to apply to the wrong overlays too eagerly.

**Comment by flokli** [comment]
Explicitly filtering on my side did workaround, so indeed filtering seems to be the problem:

```
hardware.deviceTree.filter = "rk3588-orangepi-5-plus.dtb";
```

```
Processing source device tree rockchip/rk3588-orangepi-5-plus.dtb...
  Applying overlay fan-cooling-levels
Saving final device tree rockchip/rk3588-orangepi-5-plus.dtb...
```

**Comment by K900** [comment]
This is actually intentional, because compatible matching in the old script was completely broken, it just happened to be broken in a way that made it work for your exact case. The root of the problem here is that `compatible` strings in overlays are treated as an "or", not "and", so `"xunlong,orangepi-5-plus", "rockchip,rk3588"` applies to any DT with compatible strings `"xunlong,orangepi-5-plus"` _or_ `"rockchip,rk3588"`. You want to remove the `"rockchip,rk3588"` compatible.

**Comment by flokli** [comment]
Can you send a release note update explaining this? I'm probably not the only one running into it, and this changed without any mention currently...

**Comment by K900** [comment]
https://github.com/NixOS/nixpkgs/pull/344192

**Comment by flokli** [comment]
Thanks!

**Pull request merged**